contracts/MemeBase.sol

```diff
@@ -84,61 +84,41 @@ contract MemeBase is MemeFactory {
     /// @dev MemeBase constructor
     constructor(
         address _olas,
-        address _usdc,
+        address _weth,
         address _router,
         address _factory,
         uint256 _minNativeTokenValue,
-        address _weth,
         address _l2TokenRelayer,
-        address _oracle,
         address _balancerVault,
         bytes32 _balancerPoolId
-    ) MemeFactory(_olas, _usdc, _router, _factory, _minNativeTokenValue) {
-        weth = _weth;
+    ) MemeFactory(_olas, _weth, _router, _factory, _minNativeTokenValue) {
         l2TokenRelayer = _l2TokenRelayer;
-        oracle = _oracle;
         balancerVault = _balancerVault;
         balancerPoolId = _balancerPoolId;
     }
 
-    /// @dev Buys USDC on UniswapV2 using ETH amount.
-    /// @param nativeTokenAmount Input ETH amount.
-    /// @return Stable token amount bought.
-    function _convertToReferenceToken(uint256 nativeTokenAmount, uint256) internal override returns (uint256) {
-        address[] memory path = new address[](2);
-        path[0] = weth;
-        path[1] = referenceToken;
-
-        // Calculate price by Oracle
-        (, int256 answerPrice, , , ) = IOracle(oracle).latestRoundData();
-        require(answerPrice > 0, "Oracle price is incorrect");
-
-        // Oracle returns 8 decimals, USDC has 6 decimals, need to additionally divide by 100 to account for slippage
-        // ETH: 18 decimals, USDC leftovers: 2 decimals, percentage: 2 decimals; denominator = 18 + 2 + 2 = 22
-        uint256 limit = uint256(answerPrice) * nativeTokenAmount * SLIPPAGE / 1e22;
-        // Swap ETH for USDC
-        uint256[] memory amounts = IUniswap(router).swapExactETHForTokens{ value: nativeTokenAmount }(
-            limit,
-            path,
-            address(this),
-            block.timestamp
-        );
-
-        // Return the USDC amount bought
-        return amounts[1];
+    /// @dev Get safe slippage amount from dex.
+    /// @return safe amount of tokens to swap on dex with low slippage.
+    function _getLowSlippageSafeSwapAmount() internal virtual override returns (uint256) {
+        /// check on three-sided USDC, ETH, OLAS pool for correct amount with max 3% slippage
+        return 0;
     }
 
+
     /// @dev Buys OLAS on Balancer.
-    /// @param referenceTokenAmount USDC amount.
+    /// @param nativeTokenAmount Native token amount.
     /// @param limit OLAS minimum amount depending on the desired slippage.
     /// @return Obtained OLAS amount.
-    function _buyOLAS(uint256 referenceTokenAmount, uint256 limit) internal override returns (uint256) {
-        // Approve usdc for the Balancer Vault
-        IERC20(referenceToken).approve(balancerVault, referenceTokenAmount);
+    function _buyOLAS(uint256 nativeTokenAmount, uint256 limit) internal override returns (uint256) {
+        // Wrap ETH
```

Comment:
You're wrapping twice. this already happens elsewhere

Reply:
This is correct because I wrap before the pair creation and then for OLAS exchange as those are two different amounts. But I better wrap the full stake before the pair creation.